File: problems/树/199.二叉树的右视图.py
#!/usr/bin/env python3
"""
199. 二叉树的右视图
"""
from typing import List, Optional
from collections import deque
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def right_side_view(root: Optional[TreeNode]) -> List[int]:
    if not root: return []
    result = []
    queue = deque([root])
    while queue:
        size = len(queue)
        for i in range(size):
            node = queue.popleft()
            if i == size - 1:
                result.append(node.val)
            if node.left:  queue.append(node.left)
            if node.right: queue.append(node.right)
    return result

if __name__ == "__main__":
    root1 = TreeNode(1, TreeNode(2, None, TreeNode(5)), TreeNode(3))
    print(right_side_view(root1))  # [1, 3, 5]
    print(right_side_view(TreeNode(1)))  # [1]
    print(right_side_view(None))  # []

File: problems/树/199.二叉树的右视图.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

vector<int> rightSideView(TreeNode* root) {
    if (!root) return {};
    vector<int> result;
    queue<TreeNode*> q;
    q.push(root);
    while (!q.empty()) {
        int sz = q.size();
        for (int i = 0; i < sz; i++) {
            TreeNode* node = q.front(); q.pop();
            if (i == sz - 1) result.push_back(node->val);
            if (node->left)  q.push(node->left);
            if (node->right) q.push(node->right);
        }
    }
    return result;
}
int main() {
    TreeNode* root = new TreeNode(1);
    root->left  = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->right = new TreeNode(5);
    for (int v : rightSideView(root)) cout << v << " "; cout << endl;  // 1 3 5
    for (int v : rightSideView(new TreeNode(1))) cout << v << " "; cout << endl;  // 1
    for (int v : rightSideView(nullptr)) cout << v << " "; cout << endl;  // (empty)
    return 0;
}
